build-server/src/main/java/nl/tudelft/ewi/build/builds/BuildManager.java

```diff
@@ -1,120 +1,390 @@
+			this.uuid = UUID.randomUUID();
+			this.logger = logger;
+		}
+		
+		@Override
+		public ContainerExit call() throws Exception {
+			if(!started.compareAndSet(false, true))
+				throw new IllegalStateException("DockerRunner is already running!");
+
+			File stagingDirectory = createStagingDirectory();
+			stagingDirectoryReference.set(stagingDirectory);
+			prepareStagingDirectory(stagingDirectory);
+
+			String volume = String.format("%s:%s", stagingDirectory, WORK_DIR);
+
+			BuildInstructionInterpreter<BuildInstruction> buildInstructionInterpreter =
+					getBuildIntstructionInterpreter();
+			BuildInstruction buildInstruction = buildRequest.getInstruction();
+			
+			ContainerConfig.Builder configBuilder = ContainerConfig.builder()
+					.image(buildInstructionInterpreter.getImage(buildInstruction))
+					.cmd(buildInstructionInterpreter.getCommand(buildInstruction).split(" "))
+					.volumes(volume)
+					.workingDir(WORK_DIR);
+
+			String id;
+
+			try {
+				log.info("Create container {}", config);
+				ContainerCreation creation = dockerClient.createContainer(configBuilder.build(), uuid.toString());
+				id = creation.id();
+				containerId.set(id);
+				log.info("Starting container {}", id);
+				dockerClient.startContainer(id, HostConfig.builder().binds(volume).build());
+			}
+			catch (DockerException | InterruptedException e) {
+				logger.println("[FATAL] Failed to provision build environment");
+				throw e;
+			}
+
+			try(LogStream stream = dockerClient.attachContainer(id, AttachParameter.LOGS,
+					AttachParameter.STDERR, AttachParameter.STDOUT, AttachParameter.STREAM)) {
+				log.info("Attaching log for container {}", id);
+				while(stream.hasNext() && !Thread.currentThread().isInterrupted())
+					logger.consume(stream.next());
+			}
+
+			log.info("Waiting for container to terminate {}", id);
+			return dockerClient.waitContainer(id);
+		}
+
+		/**
+		 * Kill the Docker container. This is called when the Future was cancelled
+		 * and thus the container did not exit yet.
+		 */
+		public void kill() {
+			String id = containerId.get();
+			if(id != null) {
+				log.info("Trying to kill container {}", id);
+				try {
+					dockerClient.killContainer(id);
+				}
+				catch (DockerException | InterruptedException e) {
+					log.warn("Failed to kill container " + id, e);
 				}
 			}
-		};
-	}
+		}
 
-	private Runnable createTerminator(final UUID buildId) {
-		return new Runnable() {
-			@Override
-			public void run() {
-				log.debug("Forcefully terminating build: {}", buildId);
-				killBuild(buildId);
+		/**
+		 * Remove the Docker container. This is called when the Future is either
+		 * cancelled or completed. This also removes the staging directory that
+		 * was shared with the container. 
+		 */
+		public void remove() {
+			String id = containerId.get();
+			if(id != null) {
+				log.info("Trying to remove container {}", id);
+				try {
+					dockerClient.removeContainer(id, true);
```

Comment:
This is how this branch enforces to remove the staging dir.

[`removeContainer(id, true)`](https://github.com/spotify/docker-client/blob/master/src/main/java/com/spotify/docker/client/DefaultDockerClient.java#L544) also removes the mounted volumes.

From the [Docker API reference](https://docs.docker.com/reference/api/docker_remote_api_v1.12/#remove-a-container):
> # Remove a container
> `DELETE /containers/(id)`
>
> Query Parameters:
> **v** – 1/True/true or 0/False/false, Remove the volumes associated to the container. Default false
> **force** – 1/True/true or 0/False/false, Removes the container even if it was running. Default false